CHANGES.rst

```diff
@@ -134,6 +134,20 @@
   `dolmen.builtins <https://pypi.org/project/dolmen.builtins/>`_.
   See `issue 138 <https://github.com/zopefoundation/zope.interface/issues/138>`_.
 
+- Make ``providedBy()`` and ``implementedBy()`` respect ``super``
+  objects. For instance, if class ``Derived`` implements ``IDerived``
+  and extends ``Base`` which in turn implements ``IBase``, then
+  ``providedBy(super(Derived, derived))`` will return ``[IBase]``.
+  Previously it would have returned ``[IDerived]`` (in general, it
+  would previously have returned whatever would have been returned
+  without ``super``).
+
+  Along with this change, adapter registries will unpack ``super``
+  objects into their ``__self___`` before passing it to the factory.
+  Together, this means that ``component.getAdapter(super(Derived,
+  self), ITarget)`` is now meaningful.
```

Comment:
While the unpacking is sufficient for my original use case, this may not always be the case, i.e. in some cases, one might want that the super object (and not its `__self__`) is passed to the adapter. I agree that given the current interface, it is impossible to express what the use case wants. In the long term, I would like the interface to be enhanced by a `generateAdapterFactories` decoupling adapter factory lookup and calling them. This would also significantly facilitate the implementation of "conditional adapters" (a concept, I have used for `Products.AdvancedQeury`).

Reply:
I don't have a strong immediate use case, but it seemed to me that passing the `super` object was both not very useful and inconsistent with other uses of both `super`. The practical uses I could think of for today all work best with `__self__` unwrapped.

It's inconsistent because when `super(Class, self).method()` is executing, the `self` that's sent to `method` isn't the `super` object, it's straight up exactly the `self` object provided to `super` (because `super` isn't particularly useful; see next paragraph).  Adapting on `super` ought to implicitly behave the same simply to match that expectation.

It's not particularly useful because the `super` object cannot access instance variables of `self` (only class descriptors/methods), nor (as a consequence) can it discover any interfaces that are `directlyProvidedBy` the `self` object. It has no MRO or type and `super` doesn't work on it, so you'd only be able to try this trick for exactly one level of hierarchy. The receiving factory would need to unwrap `__self__` in short order to do any of those things…and at that point, you're writing an adapter that's customized to know it's working with `super`, which destroys reusability: given `class Base` and `class Derived(Base)` the same adapter can't work for `super(Derived, self)` *and* `Base()`.

If one wanted to use the `super` object in an adapter, I think the better thing to do would be to register a multi-adapter on `(Interface-or-class, type)` (called with `component.getMultiAdapter(IAdapter, (self, Derived))`) or possibly `(Interface-or-class, super)`. The latter would work today (but not after this change), while the former would work today and after this change.